Improve Rules API

jsexauer opened the pull request:
This is an improvement to the rules API to use operators for rules:
- `&` for and
- `|` for or
- `~` for not

I've tested it with some pretty complex logic and it seems to work well.  This isn't in an end state, but I feel like enough is done to submit for PR for feedback.

JDWarner commented on skfuzzy/control/controlsystem.py:
```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+                raise RuntimeError("Antecedent already has calculated "
+                "membership.  Are you trying to computer a simulation multiple "
+                "times?  Create multiple ControlSystemSimulation objects "
+                "instead.")
```
IMO, iterative re-computation is a use case we should support without requiring new objects each time. 

Since we aren't worrying about caching at the moment, can we just overwrite these membership values and continue forward?

JDWarner commented on skfuzzy/control/controlsystem.py:
```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+        """
+        # TODO: Tracking and caching
+
+        # Check if any fuzzy variables lack input values and fuzzyfy inputs
```
minor: "fuzzify"

JDWarner commented on skfuzzy/control/controlsystem.py:
```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+        assert isinstance(ctrl, ControlSystem)
+        self.ctrl = ctrl
+        self._cache = []
+        self._cahced_graph = None
```
`_cahced_graph` -> `_cached_graph`

JDWarner commented on skfuzzy/control/controlsystem.py:
```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+
+
 
     def print_state(self):
```
This prints a great summary - let's have this get wrapped by (or become) `__repr__` so users get the info any time the object itself is passed.

jsexauer replied:
My concern with adding this to __repr__ would be that it can pretty easily generate and error (if compute hasn't been called yet, for example).  It is also really long.  Both of those indicate to me it is best not to use this for __repr__

JDWarner commented on skfuzzy/control/controlsystem.py:
```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+        if self.ctrl.consequents.next().output[self] is None:
+            raise ValueError("Call compute method first.")
+
+        print "============="
```
These print statements won't work in Python 3. The best way to harmonize behavior is to add `from __future__ import print_function` as the top import statement in any file with `print` - or just all files regardless - then convert any `print 'string'` statements to `print('string')` functions.

JDWarner replied:
Also, we should test this print behavior is working as intended - simple string comparison. The greater value is that it'll also verify the behavior works in both Python 2.x and 3.x.

JDWarner commented on skfuzzy/control/state.py:
```diff
@@ -0,0 +1,54 @@
+
+from weakref import WeakKeyDictionary
```
This is well done.

jsexauer replied:
Thanks

JDWarner commented on skfuzzy/control/state.py:
```diff
@@ -0,0 +1,54 @@
+
+    def __init__(self, initial_condition = None):
+        self.default = initial_condition
+        self.__sim_data = WeakKeyDictionary()
```
One leading underscore would probably be sufficient

JDWarner commented on skfuzzy/control/visualization.py:
```diff
@@ -48,21 +56,27 @@ def view(self, *args, **kwargs):
     def _init_plot(self):
         # Formatting: limits
-        self.ax.set_ylim([0, 1])
+        self.ax.set_ylim([0, 1.01])
```
Was matplotlib's default placing the border over data? I'm fine with this, but another approach would be simply to turn off the top (and right) axes/borders. Perhaps keep the tick marks as placeholders but nothing else.

jsexauer replied:
I'm already turnning off the axis on line 96.  This was done because if a membership is 1.0 it doesn't show up well.  In your examples, it looks fine because you're using triangle membership functions and they're 1.0 for a single point and thus you don't really notice it.  But when a mf is at 1.0 for a long time, I was noticing it was hard to see.

JDWarner commented on skfuzzy/control/visualization.py:
```diff
@@ -48,21 +56,27 @@ def view(self, *args, **kwargs):
-                self.ax.plot([crip_value] * 2, [0, y],
+                                      output_mf, crisp_value)
+                if y < 0.1:
+                    y = 1.
```
This really worried me for a moment, then I got it. Small lines don't show.

Made me think though: would it make sense to invert the line if membership is below 0.5? A vertical line starting from the top, going down to the membership value instead of up from the bottom? Or do you think that would confuse more than it would clarify?

jsexauer replied:
I think it's a cool idea, but I do think it might be confusing.  I was actually considering removing the whole adjustable line height thing and making them all a length of 1.  But I do like how they look when the value is a reasonable value above 0.1.  I added a comment, but we can re-evaluate.

JDWarner commented on skfuzzy/control/controlsystem.py:
```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+        else:
+            if len(skipped_rules) == len_rules:
+                # Avoid being caught in an infinite loop
+                raise RuntimeError("Unable to resolve rule execution order")
```
Could we provide the user with any better info about how they could fix this?

JDWarner commented on skfuzzy/control/controlsystem.py:
```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+        #  which is what we determined in step 1.  The only difference would
+        #  be if the consequent has a weight, which we would apply now.
+        for c in rule.consequent:
+            assert isinstance(c, WeightedConsequent)
```
I had to read all the changes and come back to this. Is it correct that, _every_ consequent becomes a `WeightedConsequent` with weight of `1.` provided if it was initially unweighted?

jsexauer replied:
Correct, see line 115 of rule.py (where the consequent of a rule is set).  I felt this was the most consistent way to deal with weighting consequents as prescribed by IEC 1131-7.  Figured this was a good time to include this feature while we were working in this area of the codebase.

http://www.fuzzytech.com/binaries/ieccd1.pdf page 14.

JDWarner commented on skfuzzy/control/controlsystem.py:
```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+        #   The process of actual accumulation is delegated to the
+        #   FuzzyVariableTerm which uses its parent's accumulation method
+        for c in rule.consequent:
+            assert isinstance(c, WeightedConsequent)
```
Unnecessary check as this was verified above, should be save to remove.

jsexauer replied:
assert statements help my IDE (pyCharm) understand the variable types that are being used, so I use them even when it's somewhat obvious what the variable type is.

JDWarner commented on skfuzzy/control/controlsystem.py:
```diff
@@ -329,28 +132,266 @@ def inputs(self, input_dict):
+            calced_degree = len(self.calced_graph.predecessors(p))
+            if all_degree != calced_degree:
+                return False
+        return True
```
Git changelogs are cleaner if every files ends with exactly one newline. It's because the newline is actually a hidden `\n` on the prior line, so any extensions to the file "change" the previous final line in the diff unless it was there before.

JDWarner commented on skfuzzy/control/fuzzyvariable.py:
```diff
@@ -13,8 +15,32 @@
+    pass
 
-class FuzzyVariableTerm(object):
+    def membership_value(self):
```
Should we still expose this here at all, if it's not implemented? Or is this an idea to be built on.

jsexauer replied:
While not strictly needed in python, I usually add these stubs to abstract members of abstract classes to make it more clear to anyone who does a future extension of the class what they should be implementing in the concrete class.  Another way would be to use the `abstractpropery` decorator in the `abc` module, but I think this also gets the point across without using the `abc` metaclass.

JDWarner commented on skfuzzy/control/fuzzyvariable.py:
```diff
@@ -13,8 +15,32 @@
 except ImportError:
     from .ordereddict import OrderedDict
 
+class TermPrimitive(object):
```
Nicely done.

JDWarner commented:
I like the approach a great deal. Made a few comments above, which are generally minor.

If there are any aspects/improvements you'd like to delegate, let me know! Happy to make PRs against this branch, I just don't want to duplicate effort or make messy rebases.

jsexauer commented:
Sorry, I'm a little slow on the weekdays sometimes to responding to github issues.  I've commented on or addressed all your comments.  I think this is in a stable enough state to be pulled in.  Some unit tests still aren't passing and a lot more testing and documentation needs to be written.  I will try to get to that stuff this week.

JDWarner commented:
If there are certain aspects of this you'd like to delegate, let me know - I'm happy to help but don't want to duplicate work or make messy rebases.
